core/indigo-core/common/math/algebra.h

```diff
@@ -822,5 +824,234 @@ namespace indigo
+        float area = convexPolygonArea(poly1);
+        if (area == 0.0f)
+            return 0.0f;
+        std::vector<Vec2f> intersection = convexClip(poly1, poly2);
+        if (intersection.empty())
+            return 0.0f;
+        float intersectionArea = convexPolygonArea(intersection);
+        if (std::abs(intersectionArea - area) < 1e-6f)
+            return 1.0f;
+        return intersectionArea / area;
+    }
+
+    inline float distancePointToEdge(const Vec2f& p, const Vec2f& a, const Vec2f& b)
+    {
+        Vec2f ab = b - a;
+        Vec2f ap = p - a;
+        float t = Vec2f::dot(ap, ab) / Vec2f::dot(ab, ab);
+        t = std::max(0.0f, std::min(1.0f, t));
+        Vec2f projection(a.x + ab.x * t, a.y + ab.y * t);
+        Vec2f diff = p - projection;
+        return diff.length();
+    }
+
+    inline float computeConvexDistance(const std::vector<Vec2f>& poly1, const std::vector<Vec2f>& poly2)
+    {
+        float minDist = std::numeric_limits<float>::max();
+        for (const auto& p : poly1)
+        {
+            float dist = std::numeric_limits<float>::max();
+            size_t n = poly2.size();
+            for (size_t i = 0; i < n; ++i)
+            {
+                float d = distancePointToEdge(p, poly2[i], poly2[(i + 1) % n]);
+                dist = std::min(dist, d);
+                if (d < minDist)
+                    break;
+            }
+            minDist = std::min(minDist, dist);
+        }
+        for (const auto& p : poly2)
+        {
+            float dist = std::numeric_limits<float>::max();
+            size_t n = poly1.size();
+            for (size_t i = 0; i < n; ++i)
+            {
+                float d = distancePointToEdge(p, poly1[i], poly1[(i + 1) % n]);
+                dist = std::min(dist, d);
+                if (d < minDist)
+                    break;
+            }
+            minDist = std::min(minDist, dist);
+        }
+        return minDist;
+    }
+
+    inline bool doesVerticalLineIntersectPolygon(float x, const std::vector<Vec2f>& poly)
+    {
+        for (size_t i = 0, n = poly.size(); i < n; ++i)
+        {
+            const Vec2f& a = poly[i];
+            const Vec2f& b = poly[(i + 1) % n];
+            if ((x > std::min(a.x, b.x) && x < std::max(a.x, b.x)))
+                return true;
+        }
+        return false;
+    }
+
+    inline bool doesHorizontalLineIntersectPolygon(float y, const std::vector<Vec2f>& poly)
+    {
+        for (size_t i = 0, n = poly.size(); i < n; ++i)
+        {
+            const Vec2f& a = poly[i];
+            const Vec2f& b = poly[(i + 1) % n];
+            if ((y > std::min(a.y, b.y) && y < std::max(a.y, b.y)))
+                return true;
+        }
+        return false;
+    }
+
+    inline bool doesRayIntersectPolygon(const Vec2f& p1, const Vec2f& p2, const std::vector<Vec2f>& poly)
+    {
+        Vec2f dir = p2 - p1;
+        for (size_t i = 0, n = poly.size(); i < n; ++i)
+        {
+            Vec2f A = poly[i];
```

Comment:
I think it should be "const Vect2f&" or simple "vect2f a=poly[i] - p1"

Reply:
fixed